### PR TITLE
Implements basic DHCPv6 support

### PR DESCRIPTION
**Pull request by bnaecker**
- Adds core support for DHCPv6 types, including DUIDs, Options, and the various message types.
- Adds an OPTE API action for describing DHCPv6 emulation in OPTE, with the addresses to be leased, their lifetimes, and a few other options.
- Implements the DHCPv6 protocol inside the `Dhcpv6Action::gen_packet` method.
- Adds helper implementations of `AsRef` and `Deref` to MAC and IPv6 addresses, for referring to those as byte arrays without an additional copy.
- Adds ULP checksum computation to IPv6, and fixes small edge-case with computed checksums of zero.
- Adds a bunch of units for DHCPv6, mostly for serialization/deserialization, including test data from a packet capture.
- Adds integration tests verifying the core DHCPv6 protocol for leasing addresses, including Solicit -> Advertise -> Request -> Reply, and Solicit -> Reply.

### CODE REVIEW

**Comment by bnaecker** [comment]
I've done a good bit of testing, though I'd still like to verify one or two more guest types before merging this.

For more testing notes, here's the result of starting the DHCPv6 client on a Debian guest:

```bash
debian@myinst:~$ ip -6 addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 state UNKNOWN qlen 1000
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp0s8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000
    inet6 fe80::aa40:25ff:fefd:42cb/64 scope link
       valid_lft forever preferred_lft forever
debian@myinst:~$ sudo /sbin/dhclient -6 enp0s8
debian@myinst:~$ ip -6 addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 state UNKNOWN qlen 1000
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp0s8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000
    inet6 fd00:abcd:abcd:abcd::5/128 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::aa40:25ff:fefd:42cb/64 scope link
       valid_lft forever preferred_lft forever
debian@myinst:~$
```

Here's a packet capture from the VNIC provided to the guest zone:

```bash
bnaecker@feldspar : ~/opte/xde $ pfexec snoop -i dhcpv6-3.snoop
  1   0.00000 fe80::aa40:25ff:fefd:42cb -> ff02::1:2    DHCPv6 Solicit xid=b20142 IAs=1
  2   0.00007 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fefd:42cb DHCPv6 Advertise xid=b20142 IAs=1
  3   1.04018 fe80::aa40:25ff:fefd:42cb -> ff02::1:2    DHCPv6 Request xid=c922d0 IAs=1
  4   0.00013 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fefd:42cb DHCPv6 Reply xid=c922d0 IAs=1
bnaecker@feldspar : ~/opte/xde $
```

The cool part is the last reply, which provided a few (hard-coded) DNS servers to the guest:

```bash
bnaecker@feldspar : ~/opte/xde $ pfexec snoop -i dhcpv6-3.snoop -v

<SNIP>

DHCPv6: ----- Dynamic Host Configuration Protocol Version 6 -----
DHCPv6:
DHCPv6: Message type (msg-type) = 7 (Reply)
DHCPv6: Transaction ID = c922d0
DHCPv6:
DHCPv6: Option Code = 2 (Server Identifier)
DHCPv6:   DUID Type = 3 (Link-layer Address)
DHCPv6:   Hardware Type = 1 (Ethernet (10Mb))
DHCPv6:   Link Layer Address = a8:40:25:ff:77:77
DHCPv6: Option Code = 1 (Client Identifier)
DHCPv6:   DUID Type = 1 (Link-layer Address Plus Time)
DHCPv6:   Hardware Type = 1 (Ethernet (10Mb))
DHCPv6:   Time = 718171097 (Tue Oct  4 03:58:17 2022)
DHCPv6:   Link Layer Address = a8:40:25:fd:42:cb
DHCPv6: Option Code = 23 (DNS Recursive Name Server)
DHCPv6:   Address = 2606:4700:4700::1111 (one.one.one.one)
DHCPv6:   Address = 2606:4700:4700::1001 (one.one.one.one)
DHCPv6:   Address = 2001:4860:4860::8888 (dns.google)
DHCPv6:   Address = 2001:4860:4860::8844 (dns.google)
DHCPv6: Option Code = 3 (Identity Association for Non-temporary Addresses)
DHCPv6:   IAID = 637354699
DHCPv6:   T1 (renew) = 4294967295 seconds
DHCPv6:   T2 (rebind) = 4294967295 seconds
DHCPv6: IA: ----- Identity Association -----
DHCPv6: IA: Option Code = 5 (IA Address)
DHCPv6: IA:   Address = fd00:abcd:abcd:abcd::5
DHCPv6: IA:   Preferred lifetime = 4294967295 seconds
DHCPv6: IA:   Valid lifetime = 4294967295 seconds
DHCPv6:
```

We can then see the guest start to try to use those DNS servers, so things seem to be working pretty well as far as communicating addresses and options to the guest. Here are a couple of lines snipped from the output of the `opte-port-process` D script:

```
opte0        out 3        UDP,[fd00:abcd:abcd:abcd::5]:60580,[2001:4860:4860::8888]:53 UDP,[fd00:abcd:abcd:abcd::5]:60580,[2001:4860:4860::8888]:53 101   Drop { reason: Layer { name: "router" } }
opte0        out 3        UDP,[fd00:abcd:abcd:abcd::5]:60580,[2001:4860:4860::8888]:53 UDP,[fd00:abcd:abcd:abcd::5]:60580,[2001:4860:4860::8888]:53 101   Drop { reason: Layer { name: "router" } }
```

That source address is junk I've provided in my modified Omicron setup. The destination is one of the hard-coded DNS servers.

**Comment by bnaecker** [comment]
This should resolve #245, though the full protocol is not completely implemented. I'd prefer to get this in, and implement the rest if / when we find that guests need it.

**Comment by bnaecker** [comment]
Testing with Ubuntu, and things are less sanguine that Debian. NDP seems to configure just fine, and the instance then immediately starts DHCPv6 (since the managed flag is set by OPTE's NDP router advertisements, presumably). The Solicit messages are received and hairpinned back by OPTE just fine, but the guest doesn't seem to care for them. It sends several in short succession:

```bash
bnaecker@feldspar : ~/opte/xde $ pfexec snoop -i dhcpv6.snoop
  1   0.00000 fe80::aa40:25ff:fef3:a1a7 -> ff02::1:2    DHCPv6 Solicit xid=91e930 IAs=1
  2   0.00005 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fef3:a1a7 DHCPv6 Reply xid=91e930 IAs=1
  3   1.04438 fe80::aa40:25ff:fef3:a1a7 -> ff02::1:2    DHCPv6 Solicit xid=91e930 IAs=1
  4   0.00009 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fef3:a1a7 DHCPv6 Reply xid=91e930 IAs=1
  5   2.03954 fe80::aa40:25ff:fef3:a1a7 -> ff02::1:2    DHCPv6 Solicit xid=91e930 IAs=1
  6   0.00008 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fef3:a1a7 DHCPv6 Reply xid=91e930 IAs=1
  7   3.88997 fe80::aa40:25ff:fef3:a1a7 -> ff02::1:2    DHCPv6 Solicit xid=91e930 IAs=1
  8   0.00010 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fef3:a1a7 DHCPv6 Reply xid=91e930 IAs=1
  9   7.65022 fe80::aa40:25ff:fef3:a1a7 -> ff02::1:2    DHCPv6 Solicit xid=91e930 IAs=1
 10   0.00006 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fef3:a1a7 DHCPv6 Reply xid=91e930 IAs=1
bnaecker@feldspar : ~/opte/xde $
```

Systemd appears to be unhappy with those. If I manually start ISC's DHCP client, it starts up the a 4-message exchange, and finds the address just fine:

```
ubuntu@myinst:~$ sudo /sbin/dhclient -6 enp0s8
ubuntu@myinst:~$ ip -6 addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 state UNKNOWN qlen 1000
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp0s8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000
    inet6 fd00:abcd:abcd:abcd::5/128 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::aa40:25ff:fef3:a1a7/64 scope link
       valid_lft forever preferred_lft forever
ubuntu@myinst:~$
```

I might dig into systemd's DHCP client implementation, since Ubuntu is probably going to be pretty widely-used.

**Comment by bnaecker** [comment]
Not sure systemd is the problem here anymore. It looks pretty happy from the output of `systemctl status systemd-networkd`. No error messages, nothing about even starting up DHCPv6. I'm really not sure who's sending out those initial Solicits, only to ignore them. I'll dig some more.

**Comment by rzezeski** [comment]
> Testing with Ubuntu, and things are less sanguine that Debian. NDP seems to configure just fine, and the instance then immediately starts DHCPv6 (since the managed flag is set by OPTE's NDP router advertisements, presumably). The Solicit messages are received and hairpinned back by OPTE just fine, but the guest doesn't seem to care for them. It sends several in short succession:
> 
> ```shell
> bnaecker@feldspar : ~/opte/xde $ pfexec snoop -i dhcpv6.snoop
>   1   0.00000 fe80::aa40:25ff:fef3:a1a7 -> ff02::1:2    DHCPv6 Solicit xid=91e930 IAs=1
>   2   0.00005 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fef3:a1a7 DHCPv6 Reply xid=91e930 IAs=1
>   3   1.04438 fe80::aa40:25ff:fef3:a1a7 -> ff02::1:2    DHCPv6 Solicit xid=91e930 IAs=1
>   4   0.00009 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fef3:a1a7 DHCPv6 Reply xid=91e930 IAs=1
>   5   2.03954 fe80::aa40:25ff:fef3:a1a7 -> ff02::1:2    DHCPv6 Solicit xid=91e930 IAs=1
>   6   0.00008 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fef3:a1a7 DHCPv6 Reply xid=91e930 IAs=1
>   7   3.88997 fe80::aa40:25ff:fef3:a1a7 -> ff02::1:2    DHCPv6 Solicit xid=91e930 IAs=1
>   8   0.00010 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fef3:a1a7 DHCPv6 Reply xid=91e930 IAs=1
>   9   7.65022 fe80::aa40:25ff:fef3:a1a7 -> ff02::1:2    DHCPv6 Solicit xid=91e930 IAs=1
>  10   0.00006 fe80::aa40:25ff:feff:7777 -> fe80::aa40:25ff:fef3:a1a7 DHCPv6 Reply xid=91e930 IAs=1
> bnaecker@feldspar : ~/opte/xde $
> ```
> 
> Note that these are the 2-message exchange, Solicit + Rapid Commit -> Reply. I have an integration test verifying this behavior, and the Reply itself looks OK. Still, systemd appears to be unhappy with those. If I manually start ISC's DHCP client, it starts up the a 4-message exchange, and finds the address just fine:
> 
> ```
> ubuntu@myinst:~$ sudo /sbin/dhclient -6 enp0s8
> ubuntu@myinst:~$ ip -6 addr
> 1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 state UNKNOWN qlen 1000
>     inet6 ::1/128 scope host
>        valid_lft forever preferred_lft forever
> 2: enp0s8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP qlen 1000
>     inet6 fd00:abcd:abcd:abcd::5/128 scope global
>        valid_lft forever preferred_lft forever
>     inet6 fe80::aa40:25ff:fef3:a1a7/64 scope link
>        valid_lft forever preferred_lft forever
> ubuntu@myinst:~$
> ```
> 
> I might dig into systemd's DHCP client implementation, since Ubuntu is probably going to be pretty widely-used.

When you say "systemd appears to be unhappy with those", does that mean you see some type of error?

Does `netstat -s` (in the guest) report non-zero for any error counts? If you get a pcap of the guest traffic (either from the VNIC or from inside the guest) and open it in wireshark does it report any issues with the packets?

Is there a reason the systemd-induced DHCP sequence is attempting to use a rapid-commit sequence and the manual invocation of DHCPv6 is not? Are they two completely different implementations? (Oh right, I guess the point of systemd is to replace all existing software). If you look at the DHCPv6 message in wireshark for the rapid-commit sequence do they include the correct options and conform to the requirements laid out in RFC 8415?

**Comment by bnaecker** [comment]
> When you say "systemd appears to be unhappy with those", does that mean you see some type of error?

Yeah, that wasn't clear. No, there are no errors, I just see repeated Solicit messages. They're the same message repeated, and obviously `xde` is hairpinning the same Reply.

> Does netstat -s (in the guest) report non-zero for any error counts? If you get a pcap of the guest traffic (either from the VNIC or from inside the guest) and open it in wireshark does it report any issues with the packets?

I'll look at `netstat` now. I have a pcap, and everything looks OK. Wireshark doesn't flag any errors in checksums, invalid data, or anything obvious. I've looked at those packets manually, and it seems OK as far as I can tell. I can share the pcap if that's useful.

> Is there a reason the systemd-induced DHCP sequence is attempting to use a rapid-commit sequence and the manual invocation of DHCPv6 is not? Are they two completely different implementations? (Oh right, I guess the point of systemd is to replace all existing software). If you look at the DHCPv6 message in wireshark for the rapid-commit sequence do they include the correct options and conform to the requirements laid out in RFC 8415?

Yeah, these are two separate userland DHCPv6 clients, `systemd-networkd` and the ISC's `dhclient`. As far as I can see, the Solicit + Rapid Commit from `systemd` appears to conform to RFC 8415. (I'm going to dig into it again to see what I'm mising.) They are not missing any required options, and there are no options that we can't either ignore or handle appropriately. Also, almost the exact same Reply is sent back from `xde` in response to both exchanges. (That's the first message from us in the case of `systemd` and the second in the case of `dhclient`.) They differ in the client ID and transaction ID, but the actual _data_ we send is identical (DNS servers and IPv6 address).

**Comment by bnaecker** [comment]
I might have found part of the problem:

>    A server MUST include this option in a Reply message sent in response
   to a Solicit message when completing the Solicit/Reply message
   exchange.

That's from the section on Rapid Commit. I missed that the first 12 times through the RFC, but it looks like we need to send that option back in the Reply. Presumably whatever client is running is discarding the Reply as it does not have that option. I'll fix that up now, and see what happens.

File: opte/src/engine/checksum.rs [hunk 1]
```diff
@@ -59,10 +59,32 @@ impl Checksum {
...
         }
 
-        ((self.inner & 0xFFFF) as u16).to_ne_bytes()
+        let sum = (self.inner & 0xFFFF) as u16;
+        if sum == 0 {
+            [0xFF, 0xFF]
+        } else {
+            sum.to_ne_bytes()
+        }
     }
```
Comment (rzezeski) [inline opte/src/engine/checksum.rs:68]: This change is incorrect as it's happening at the wrong time. You'll never trigger this case, because a rolling checksum of all zeros is only possible if the header is all zero bytes.

The `Checksum` type is used to store a rolling one's complement checksum calculation, it **is not** the checksum value stored in the header. The `HeaderChecksum` type is the type that is actually stored/written to a header (I have a todo in my local branch to document these, but I should probably float an individual PR for it). If you look at its `From` impl you'll see it performs bitwise negation (one's complement) of the current sum, which is the final step. It's **that value** that needs the check against `0x0000`, which is a totally valid result for a TCP checksum, but results in ambiguity for a UDP checksum.

I think we should apply this special case only to the UDP code after it computes its `HeaderChecksum`, and this should not exist in the generic checksum code.

**Review by rzezeski — COMMENTED** [review]
Thanks for taking on this work, looks like it was a lot. I still have more to go through but wanted to get my checksum comment in now in case it has anything to do with the issues you are seeing.

File: opte-api/src/ip.rs [hunk 2]
```diff
@@ -478,6 +480,19 @@ impl Debug for Ipv4Addr {
     }
 }
 
+impl AsRef<[u8]> for Ipv4Addr {
+    fn as_ref(&self) -> &[u8] {
+        &self.inner
+    }
+}
+
+impl Deref for Ipv4Addr {
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply opte-api/src/ip.rs:489]: Yeah, I waffled on this too. I'm not sure I completely understand the logic here. Let's take this as an example, `Ipv4Addr` and `&[u8]`. My rough interpretation is that `AsRef` allows you to pass a `&Ipv4Addr` where a `&[u8]` is expected. The `Deref` trait instead allows us to call methods on `Ipv4Addr` that are defined by `&[u8]`. I believe we can accomplish with just `AsRef`, as long as we are happy calling slice methods with something like `addr.as_ref().slice_method()`. 

I ultimately decided this approach based on the desire to just consider addresses as byte arrays. From that lens, while I don't know if they qualify as smart pointers, it seems like they can reasonably be considered to have all the methods defined on `&[u8]`. I realize there's some taste involved here too, so I'm happy finding an alternate approach.

File: opte/src/engine/checksum.rs (hunk 1 above)
Comment (bnaecker) [reply opte/src/engine/checksum.rs:68]: Ah, I didn't realize the relationship there, or that TCP checksums of 0 are fine. I'll just revert all this now, in favor of a new PR. Thanks.

**Comment by bnaecker** [comment]
Update on the Rapid Commit issue. That indeed seemed to be the problem. Adding that option into the Reply to a Solicit + Rapid Commit message enables an Ubuntu guest to pick up the VPC-private IPv6 address in a single 2-message exchange. I've updated the tests to assert this too.

File: opte-api/src/dhcpv6.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    ///
+    /// The preferred lifetime must be no longer than the valid lifetime.
+    pub fn new(
+        addr: Ipv6Addr,
+        preferred: u32,
+        valid: u32,
+    ) -> Result<Self, String> {
+        if valid < preferred {
+            return Err(String::from(
+                "Preferred lifetime must be >= valid lifetime",
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply opte-api/src/dhcpv6.rs:90]: Yes it should thanks.

File: oxide-vpc/src/engine/dhcpv6.rs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        client_mac: cfg.private_mac,
+        server_mac: cfg.gateway_mac,
+        addrs,
+        dns_servers: vec![
+            // CloudFlare
+            Ipv6Addr::from_const([0x2606, 0x4700, 0x4700, 0, 0, 0, 0, 0x1111]),
+            Ipv6Addr::from_const([0x2606, 0x4700, 0x4700, 0, 0, 0, 0, 0x1001]),
+            // Google
+            Ipv6Addr::from_const([0x2001, 0x4860, 0x4860, 0, 0, 0, 0, 0x8888]),
+            Ipv6Addr::from_const([0x2001, 0x4860, 0x4860, 0, 0, 0, 0, 0x8844]),
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply oxide-vpc/src/engine/dhcpv6.rs:60]: Thanks, I agree. We don't have any APIs for setting or managing these in Nexus yet, so it'll be a while unfortunately. Although we could get part-way there by hard-coding them in Nexus and making them part of the `VpcCfg` object sooner.

File: oxide-vpc/src/engine/dhcpv6.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+pub fn setup(
+    pb: &mut PortBuilder,
+    cfg: &VpcCfg,
+    ft_limit: NonZeroU32,
+) -> Result<(), OpteError> {
+    // This layer contains no actions if the client has not been configured with
+    // IPv6 support.
+    let ip_cfg = match cfg.ipv6_cfg() {
+        None => return Ok(()),
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply oxide-vpc/src/engine/dhcpv6.rs:40]: I agree with the larger point. I was trying to say that _this layer_, `"dhcpv6"` has nothing to do. I could have made that clearer though. I like the idea of a first "gateway" layer that drops all outbound IPv6 traffic (or whatever it is, based on the configuration). That could do the same thing inbound, or we could have a similar layer that is first on the inbound direction. It does keep coming up though, we should at least write something up.

File: oxide-vpc/tests/integration_tests.rs [hunk 6]
```diff
@@ -2866,3 +2869,260 @@ fn test_gateway_neighbor_advert_reply() {
...
+//
+// A Solicit message normally generates an Advertise message. But if the Solicit
+// message also contains the Rapid Commit option, the server is supposed to
+// respond with a Reply instead.
+//
+// In both cases, the contained data is the same. (That's so that the client
+// could use the data from more than one server to decide which one to actually
+// make a subsequent Request to.)
+#[test]
+fn test_reply_to_dhcpv6_solicit_or_reply() {
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply oxide-vpc/tests/integration_tests.rs:3001]: Yep, thanks.

File: oxide-vpc/src/engine/dhcpv6.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bnaecker) [reply oxide-vpc/src/engine/dhcpv6.rs:40]: I meant to ask. Would you prefer I add rules here that explicitly drop all DHCPv6 traffic?

[the remark this answers is not in the document]
Comment (bnaecker) [reply oxide-vpc/src/engine/dhcpv6.rs:40]: I'm planning to drop all DHCPv6 traffic, similar to ICMPv6, when there's no IPv6 at all for the client. I still agree, a blanket rule dropping any EtherType of IPv6 is better, but this will be good in the meantime.

File: oxide-vpc/src/engine/dhcpv6.rs [hunk 7]
```diff
@@ -34,13 +44,46 @@ pub fn setup(
...
+const LAYER_NAME: &'static str = "dhcpv6";
 
+fn drop_all_dhcpv6(
+    pb: &mut PortBuilder,
+    ft_limit: NonZeroU32,
+) -> Result<(), OpteError> {
+    // Predicates identifying any traffic destined for a DHCPv6 server.
+    let predicates = vec![
+        // Destined for the server multicast IP address.
+        Predicate::InnerSrcIp6(vec![
```
Comment (rzezeski) [inline oxide-vpc/src/engine/dhcpv6.rs:62]: I think this should be for the dest port?

Comment (bnaecker) [reply oxide-vpc/src/engine/dhcpv6.rs:62]: Yeah, nice catch.

**Review by rzezeski — APPROVED** [review]
Thanks for all the hard work on this! This looks good to me.

**Comment by bnaecker** [comment]
Thanks for the review @rzezeski!
